amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,549 @@
+            "GW2ANR-18C"],
+        "OSCZ": [
+            "GW1NS-4", "GW1NS-4C",
+            "GW1NSR-4", "GW1NSR-4C",
+            "GW1NSER-4C",
+            "GW1NZ-1"],
+        "OSCF": [
+            "GW1NS-2", "GW1NS-2C",
+            "GW1NSE-2C",
+            "GW1NSR-2", "GW1NSR-2C"],
+        "OSCH": [
+            "GW1N-1", "GW1N-1S",
+            "GW1NR-1"],
+        "OSCO": [
+            "GW1N-2", "GW1N-1P5", "GW1N-2B", "GW1N-1P5B",
+            "GW1NR-2", "GW1NR-2B"],
+        "OSCW": [
+            "GW2AN-18X", "GW2AN-9X"],
+    }
+
+    # Each on-chip oscillator has unique properties
+    @property
+    def _osc_props(self):
+        props = {
+            "freq": 0,
+            "div_min": 2,
+            "div_max": 128,
+            "div_step": 2
+        }
+        osc = self._oscillator
+        if osc is "OSC":
+            if "-4" in self.device:
+                props["freq"] = 210000000
+            else:
+                props["freq"] = 250000000
+        elif osc in ["OSCZ", "OSCO"]:
+            props["freq"] = 250000000
+        elif osc in ["OSCF", "OSCH"]:
+            props["freq"] = 240000000
+        elif osc is "OSCW":
+            props["freq"] = 200000000
+        else:
+            assert False
+        return props
+
+
+    @property
+    def _oscillator(self):
+        oscillators = []
+        for osc,devs in self.osc_dev_mapping.items():
+            if self.device in devs:
+                oscillators.append(osc)
+        # should this be unique?
+        if len(oscillators) != 1:
+            assert False
+        return oscillators[0]
+
+
+    ### Common templates
+
+    _common_file_templates = {
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.cst": r"""
+            // {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                IO_LOC "{{port_name}}" {{pin_name}};
+                {% for attr_name, attr_value in attrs.items() -%}
+                    IO_PORT "{{port_name}}" {{attr_name}}={{attr_value}};
+                {% endfor %}
+            {% endfor %}
+        """,
+        "{{name}}.sdc": r"""
+        // {{autogenerated}}
+        {% for net_signal,port_signal,frequency in platform.iter_clock_constraints() -%}
+            create_clock -name {{port_signal.name|tcl_escape}} -period {{1000000000/frequency}}  [get_ports {{port_signal.name|tcl_escape}}]
+        {% endfor %}
+        {{get_override("add_constraints")|default("# (add_constraints placeholder)")}}
+        """,
+    }
+
+
+    ### Apicula templates
+
+    _apicula_required_tools = ["yosys", "nextpnr-gowin", "gowin_pack"]
+
+    _apicula_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        **_common_file_templates,
```

Comment:
That's not right—you need either `.il`+`.debug.v` or just `.v`; otherwise you are wasting time by doing double work (and this can quickly add up).

Also, I don't think nextpnr uses the SDC file at all.

Reply:
Is there a default way to distinguish between the `.il`+`.debug.v` or `.v` path? Can I run the toolchain in debug mode?

Reply:
I've removed `.debug.v` generation for now, and don't generate `.sdc` for nextpnr.

Reply:
Apicula would use `.il`+`.debug.v`, Gowin would use `.v`.

Reply:
Ok, got it. Fixing this now.